# configfile specified in command line
# threads info - Sen Li:
    # /home/projects/ku-cbd/data/HoloFood/SnakeMake_Scripts/holofood_snakemake_bwa

##
# Quality-filtering
##

rule qual_filt:
    input:
        read1="{projectpath}/00-InputData/{sample}_1.fastq.gz",
        read2="{projectpath}/00-InputData/{sample}_2.fastq.gz"
    output:
        read1="{projectpath}/01-QualityFiltered/{sample}_1.fastq",
        read2="{projectpath}/01-QualityFiltered/{sample}_2.fastq",
        stats_file="{projectpath}/01-QualityFiltered/{sample}.stats"
    threads: 4
    params:
        adapter1=expand("{adapter1}", adapter1=config['adapter1']),
        adapter2=expand("{adapter2}", adapter2=config['adapter2']),
        maxns=expand("{maxns}", maxns=config['maxns']),
        minquality=expand("{minquality}", minquality=config['minquality']),
        threads=expand("{threads}", threads=config['threads'])
    run:
        import time
        import gzip
        statsfile=open(output.stats_file,"w+")
        current_time = time.strftime("%m.%d.%y %H:%M", time.localtime())
        statsfile.write("Statistic\tValue \r\n".format(current_time))

        #Get initial stats
        reads = 0
        bases = 0
        #If gzipped
        import os
        if str(input.read1).endswith('.gz'):
            with gzip.open(str(input.read1), 'rb') as read:
                for id in read:
                    seq = next(read)
                    reads += 1
                    bases += len(seq.strip())*2
                    next(read)
                    next(read)
        else:
            with open(input.read1, 'rb') as read:
                for id in read:
                    seq = next(read)
                    reads += 1
                    bases += len(seq.strip())*2
                    next(read)
                    next(read)
        statsfile.write("Input reads\t{0} ({1} bases)\r\n".format(reads,bases))
        statsfile.close()


        shell("python ../../bin/holo-qual_filt.py -i1 {input.read1} -i2 {input.read2} -o1 {output.read1} -o2 {output.read2} -a1 {params.adapter1} -a2 {params.adapter2} -maxns {params.maxns} -minq {params.minquality} -t {params.threads}")

        #Get stats after quality filtering
        reads = 0
        bases = 0
        with open(str(output.read1), 'rb') as read:
            for id in read:
                seq = next(read)
                reads += 1
                bases += len(seq.strip())
                next(read)
                next(read)

        #Print stats to stats file
        statsfile=open(str(output.stats_file),"a+")
        statsfile.write("Quality filtered reads\t{0} ({1} bases)\r\n".format(reads,bases))
        statsfile.close()


rule dup_rem_paired:
    input:
      read1="{projectpath}/01-QualityFiltered/{sample}_1.fastq",
      read2="{projectpath}/01-QualityFiltered/{sample}_2.fastq"
    output:
      dir="{projectpath}/02-DuplicatesRemoved/{sample}.merged.fastq"
    threads: 4
    params:
        separator=expand("{separator}", separator=config['separator'])
    shell:
        "python ../../bin/holo-dup_rem_paired.py -1 {input.read1} -2 {input.read2} -o {output.dir} -sep {params.separator}"



rule dup_rem_paired_repair:
    input:
      in_file="{projectpath}/02-DuplicatesRemoved/{sample}.merged.fastq",
      in_stats="{projectpath}/01-QualityFiltered/{sample}.stats"
    output:
      read1="{projectpath}/02-DuplicatesRemoved/{sample}_1.fastq",
      read2="{projectpath}/02-DuplicatesRemoved/{sample}_2.fastq",
      stats_file="{projectpath}/02-DuplicatesRemoved/{sample}.stats"
    threads: 4
    params:
        separator=expand("{separator}", separator=config['separator'])
    run:
      shell("python ../../bin/holo-dup_rem_paired_repair.py -i {input.in_file} -1 {output.read1} -2 {output.read2} -sep {params.separator}")
      shell("mv {input.in_stats} {output.stats_file}")


      #Get stats after duplicate removal
      reads = 0
      bases = 0
      with open(str(output.read1), 'rb') as read:
        for id in read:
            seq = next(read)
            reads += 1
            bases += len(seq.strip())*2
            next(read)
            next(read)

        #Print stats to stats file
        statsfile=open(str(output.stats_file),"a+")
        statsfile.write("Dereplicated reads\t{0} ({1} bases)\r\n".format(reads,bases))
        statsfile.close()



##
# Mapping to host
##

rule map_host:
    input:
        read1="{projectpath}/02-DuplicatesRemoved/{sample}_1.fastq",
        read2="{projectpath}/02-DuplicatesRemoved/{sample}_2.fastq",
        refgenome=expand("{refgenomehost}", refgenomehost=config['refgenomehost'])
    output:
        "{projectpath}/03-MappedToHost/{sample}_all.bam"
    threads: 8
    run:
      shell("python ../../bin/holo-map_host.py -1 {input.read1} -2 {input.read2} -hostrg {input.refgenome} -obam {output}")


rule map_host_split:
    input:
        refgenome=expand("{refgenomehost}", refgenomehost=config['refgenomehost']),
        all_bam="{projectpath}/03-MappedToHost/{sample}_all.bam"
    output:
        host="{projectpath}/03-MappedToHost/{sample}_host.bam",
        read1="{projectpath}/03-MappedToHost/{sample}_1.fastq",
        read2="{projectpath}/03-MappedToHost/{sample}_2.fastq"
    shell:
        """
        python ../../bin/holo-map_host_split.py -hostrg {input.refgenome} -ibam {input.all_bam} -1 {output.read1} -2 {output.read2} -obam {output.host}
        """

##
# Mapping to human
##
rule map_human:
    input:
        read1="{projectpath}/03-MappedToHost/{sample}_1.fastq",
        read2="{projectpath}/03-MappedToHost/{sample}_2.fastq",
        refgenome=expand("{refgenomehuman}", refgenomehuman=config['refgenomehuman'])
    output:
        "{projectpath}/04-MappedToHuman/{sample}_all.bam"
    threads: 8
    shell:
        """
        python ../../bin/holo-map_human.py -1 {input.read1} -2 {input.read2} -hrg {input.refgenome} -obam {output}
        """

rule map_human_split:
    input:
        refgenome=expand("{refgenomehuman}", refgenomehuman=config['refgenomehuman']),
        all_bam="{projectpath}/04-MappedToHuman/{sample}_all.bam",
        in_stats="{projectpath}/02-DuplicatesRemoved/{sample}.stats"
    output:
        read1="{projectpath}/04-MappedToHuman/{sample}_1.fastq", ## mapped
        read2="{projectpath}/04-MappedToHuman/{sample}_2.fastq", ## mapped
        stats_file="{projectpath}/04-MappedToHuman/{sample}.stats"
    run:
        shell("python ../../bin/holo-map_human_split.py -hrg {input.refgenome} -ibam {input.all_bam} -1 {output.read1} -2 {output.read2}")
        shell("mv {input.in_stats} {output.stats_file}")

        #Get stats
        reads = 0
        bases = 0
        with open(str(output.read1), 'rb') as read:
            for id in read:
                seq = next(read)
                reads += 1
                bases += len(seq.strip())*2
                next(read)
                next(read)
        #Print stats to statsfile
        statsfile=open(str(output.stats_file),"a+")
        statsfile.write("Reads after mapping to reference genome \t{0} ({1} bases)\r\n".format(reads,bases))
        statsfile.close()

print("############################ Holoflow has finished PREPROCESSING :) ############################")"
