# 08.10.20
# Metagenomics dereplication

configfile:"/home/projects/ku-cbd/people/nurher//holoflow/workflows/metagenomics/final_stats/config.yaml"

rule get_paths:
    input:
        holopath=expand("{holopath}", holopath=config['holopath']),
        logpath=expand("{logpath}", logpath=config['logpath'])



################################################################################################################
###########################################      FINAL STATISTICS     ###########################################
################################################################################################################


##
# Map MAGs to original metagenomic fastq files
##
rule mag_mapping:
    input:
        drep_bin_dir="{projectpath}/MFS_00-InputData/{group}/dereplicated_bins",
        read_dir="{projectpath}/MFS_00-InputData/{group}/metagenomic_reads"
    output:
        directory("{projectpath}/MFS_01-MAGMapping/{group}")
    params:
        threads=expand("{threads}", threads=config['threads']),
        group="{group}"
    shell:
        """
        python {rules.get_paths.input.holopath}/bin/holo-MAG_mapping.py -fq_dir {input.read_dir} -bin_dir {input.drep_bin_dir} -out_dir {output} -t {params.threads} -ID {params.group} -log {rules.get_paths.input.logpath}
        """

##
# Get MAG coverage for each sample in group
##
rule coverage:
    input:
        drep_bin_dir="{projectpath}/MFS_00-InputData/{group}/dereplicated_bins",
        bam_MAGs="{projectpath}/MFS_01-MAGMapping/{group}"
    output:
        "{projectpath}/MFS_02-MAGCoverage/{group}/{group}.coverage_byMAG.txt"
    params:
        threads=expand("{threads}", threads=config['threads']),
        out_dir="{projectpath}/MFS_02-MAGCoverage",
        group="{group}"
    shell:
        """
        python {rules.get_paths.input.holopath}/bin/holo-MAG_coverage.py -bam_dir {input.bam_MAGs} -mag_dir {input.drep_bin_dir}  -out_dir {params.out_dir} -t {params.threads} -ID {params.group} -log {rules.get_paths.input.logpath}
        """
